File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/fs/file-presence.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-
import type {
  Finding,
  HarnessCheckRule,
  RuleContext,
} from "../harness-check-rule";

/**
 * Require specified files to exist.
 */
export const filePresenceRule: HarnessCheckRule = {
  category: "filePresence",
  applies(ctx: RuleContext): boolean {
    const section = ctx.manifest.raw.filePresence;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    if (enabled === false) {
      return false;
    }
    const entry = ctx.readJsonObject(
      (section as Record<string, unknown>).parameters,
    );
    return ctx.readStringArray(entry.paths).length > 0;
  },

  validate(ctx: RuleContext): readonly Finding[] {
    const parameters = ctx.readJsonObject(
      ctx.readJsonObject(ctx.manifest.raw.filePresence).parameters,
    );
    const paths = ctx.readStringArray(parameters.paths);
    const severity = ctx.severityOf("filePresence");
    const category = "filePresence";
    return paths.flatMap((path) => {
      if (ctx.isSymlink(path) && ctx.allowedRootContractTarget(path) === null) {
        return [
          {
            severity,
            category,
            message: `symlink file is not allowed: ${path}`,
            file: path,
            startLine: 1,
            startColumn: 1,
            endLine: 1,
            endColumn: 1,
          },
        ];
      }
      if (ctx.isFile(path)) {
        return [];
      }
      return [
        {
          severity,
          category,
          message: `missing file: ${path}`,
          file: path,
          startLine: 1,
          startColumn: 1,
          endLine: 1,
          endColumn: 1,
        },
      ];
    });
  },
};

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/fs/file_presence.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""
Require files exist rule.
"""

import sys


from collections.abc import Iterable

from rules.harness_check_rule import (
    Finding,
    FindingFix,
    FixSafety,
    HarnessCheckRule,
)
from core.rule_context import RuleContext

sys.stdout.reconfigure(encoding="utf-8")
sys.stderr.reconfigure(encoding="utf-8")


class FilePresenceRule(HarnessCheckRule):
    """Validate filePresence check."""

    category = "filePresence"

    def applies(self, ctx: RuleContext) -> bool:
        """Check if this rule applies to the context."""
        section = ctx.manifest.raw.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, ctx: RuleContext) -> Iterable[Finding]:
        """Validate filePresence check."""
        section = ctx.manifest.raw.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        paths = params.get("paths", [])
        if not isinstance(paths, list):
            return []
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        return [
            Finding(
                ctx.severity_of(self.category),
                self.category,
                messages.get("default", "missing file: {path}").format(path=path),
                file=path,
                start_line=1,
                start_column=1,
                end_line=1,
                end_column=1,
                fix=FindingFix(
                    description=f"create {path}",
                    safety=FixSafety.MANUAL,
                ),
            )
            for path in paths
            if isinstance(path, str) and not ctx.is_file(path)
        ]


RULE: HarnessCheckRule = FilePresenceRule()
